smaug/operators/padding_op.h

```diff
@@ -19,95 +20,77 @@ namespace smaug {
-      /*
-      copyTensorRegion(Tensor* dest,
-                      Tensor* src,
-                      const std::vector<int>& destOrigin,
-                      const std::vector<int>& srcOrigin,
-                      const std::vector<int>& regionSize
-      */
-      std::vector<int> destOrigin;
-      if (input->getShape().getLayout() == DataLayout::NCHW){
-        destOrigin = std::vector<int>({0, 0, padder, padder});
-      }
-      else if(input->getShape().getLayout() == DataLayout::NHWC){
-        destOrigin = std::vector<int>({0, padder, padder, 0});
-      }
-      else{
-        assert(false && "Invalid padding data type!");
-      }
-      std::vector<int> srcOrigin = std::vector<int>({0, 0, 0, 0});
-      std::vector<int> regionSize = inputDims;
-      copyTensorRegion(output, input, destOrigin, srcOrigin, regionSize);
+        Tensor* input = getInput(0);
+        Tensor* output = getOutput(0);
+        int ndims = input->ndims();
+        const std::vector<int> inputDims = input->getShape().dims();
+        const std::vector<int> outputDims = output->getShape().dims();
+        int total_dim = 1;
+        for (int i : outputDims) {
+            total_dim *= i;
+        }
+        std::vector<float> vf(total_dim, 0);
+        output->fillData(vf.data(), vf.size());
+        std::vector<int> destOrigin, paddingBegin, srcOrigin;
+        for (int i = 0; i < ndims; i++) {
+            paddingBegin.push_back(paddingSize[2 * i]);
+            srcOrigin.push_back(0);
+        }
+        destOrigin = std::vector<int>(paddingBegin);
+        std::vector<int> regionSize = inputDims;
+        copyTensorRegion(output, input, destOrigin, srcOrigin, regionSize);
     }
 
     // Optional override for testing purposes.
     void createAllTensors() override {
         Tensor* input = getInput(0);
+        int ndims = input->ndims();
         std::vector<int> dims = input->getShape().dims();
-        if (input->getShape().getLayout() == DataLayout::NCHW){
-          dims[2] += 2*padder;
-          dims[3] += 2*padder;
-        }
-        else if (input->getShape().getLayout() == DataLayout::NHWC){
-          dims[1] += 2*padder;
-          dims[2] += 2*padder;
+        for (int i = 0; i < ndims; i++) {
+            dims[i] += (paddingSize[2 * i] + paddingSize[2 * i + 1]);
         }
         TensorShape shape(
                 dims, input->getShape().getLayout(), Backend::Alignment);
         Tensor* output = new Tensor(name, shape);
         workspace->addTensor(output);
         outputs.at(0) = output;
-     }
+    }
 
     // Optional but recommended function to verify operator parameters.
     bool validate() override {
-      if (padder < 0){
-        return false;
-      }
+        Tensor* input = getInput(0);
+        int ndims = input->ndims();
+        if (paddingSize.size() != 2 * ndims) {
+            return false;
+        }
         return Operator::validate();
     }
-    
+
     enum { kInputs, kNumInputs };
```

Comment:
nit: `kInput` instead of `kInputs` (you only have one), and likewise for outputs.